AppCenter/AppCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -7,67 +7,116 @@
+  NSMutableArray *tableQueries = [NSMutableArray new];
+  
+  // Browse tables.
+  for (NSString *tableName in schema) {
 
-      // Compute table query.
-      if (tableDontExist) {
-        [tableQueries addObject:[NSString stringWithFormat:@"CREATE TABLE \"%@\" (%@);", tableName,
-                                                           [columnQueries componentsJoinedByString:@", "]]];
-      }
-      [dbColumnsIndexes setObject:tableColumnsIndexes forKey:tableName];
+    // Optimization, don't even compute the query if the table already exists.
+    if ([self tableExists:tableName inDatabase:db]) {
+      continue;
     }
-
-    // Create the DB.
-    if (tableQueries.count > 0) {
-      NSString *createTablesQuery = [tableQueries componentsJoinedByString:@"; "];
-      [self executeNonSelectionQuery:createTablesQuery];
+    NSMutableArray *columnQueries = [NSMutableArray new];
+    NSArray<NSDictionary *> *columns = schema[tableName];
+
+    // Browse columns.
+    for (NSUInteger i = 0; i < columns.count; i++) {
+      NSString *columnName = columns[i].allKeys[0];
+      
+      // Compute column query.
+      [columnQueries addObject:[NSString stringWithFormat:@"\"%@\" %@", columnName,
+                                [columns[i][columnName] componentsJoinedByString:@" "]]];
     }
+    
+    // Compute table query.
+    [tableQueries addObject:[NSString stringWithFormat:@"CREATE TABLE \"%@\" (%@);", tableName,
+                             [columnQueries componentsJoinedByString:@", "]]];
+  }
+  
+  // Create the tables.
+  if (tableQueries.count > 0) {
+    NSString *createTablesQuery = [tableQueries componentsJoinedByString:@"; "];
+    [self executeNonSelectionQuery:createTablesQuery inDatabase:db];
+  }
+  return tableQueries.count;
+}
 
-    // Keep collected indexes.
-    _columnIndexes = dbColumnsIndexes;
++ (NSDictionary *)columnsIndexes:(MSDBSchema *)schema {
+  NSMutableDictionary *dbColumnsIndexes = [NSMutableDictionary new];
+  for (NSString *tableName in schema) {
+    NSMutableDictionary *tableColumnsIndexes = [NSMutableDictionary new];
+    NSArray<NSDictionary *> *columns = schema[tableName];
+    for (NSUInteger i = 0; i < columns.count; i++) {
+      NSString *columnName = columns[i].allKeys[0];
+      [tableColumnsIndexes setObject:@(i) forKey:columnName];
+    }
+    [dbColumnsIndexes setObject:tableColumnsIndexes forKey:tableName];
   }
-  return self;
+  return dbColumnsIndexes;
 }
 
-- (BOOL)tableExists:(NSString *)tableName {
++ (BOOL)tableExists:(NSString *)tableName inDatabase:(void *)db {
   NSString *query = [NSString
       stringWithFormat:@"SELECT COUNT(*) FROM \"sqlite_master\" WHERE \"type\"='table' AND \"name\"='%@';", tableName];
-  NSArray *result = [self executeSelectionQuery:query];
+  NSArray *result = [self executeSelectionQuery:query inDatabase:db];
   return (result.count > 0) ? [result[0][0] boolValue] : NO;
 }
 
++ (NSUInteger)versionInDatabase:(void *)db {
```

Comment:
Please rename `inDatabase` by `inOpenedDatabase`, we want to make it explicit that the database has to be opened before it's passed to this method.